enhancements/microshift/microshift-pluggable-core-components.md

```diff
@@ -0,0 +1,348 @@
+alternative to this would be to use upgrade-binaries optional in components
+which are ran when necessary (similar to how CNIs work).
+
+#### Configuring
+
+Configuration of the optional components is done by package selection
+when the images are created.
+
+Specific component configuration could be handled as yaml files in
+`/etc/microshift/components.d/` . i.e.:
+
+```
+/etc/microshift/component.d/010-storage-topolvm/config.yaml:
+a:
+  b: 1234
+  c: 5677
+
+y: "abcd"
+```
+
+Would be rendered in the templates with:
+
+```
+{{ .config.a.b || 'default' }} {{ .config.a.c }} {{ config.y }}
+```
+
+Additionally more general MicroShift configuration and system details are passed
+down as variables to the templates:
+
+* `{{ .NodeName }}`
+* `{{ .NodeIP }}`
+* `{{ .ClusterCIDR }}`
+* `{{ .ServiceCIDR }}`
+* `{{ .ClusterDNS }}`
+* `{{ .BaseDomain }}`
+
+#### Deploying Applications
+
+Applications are deployed as usual, we would add support for multiple
+directories in the user component manifests as described in the 2nd goal.
+
+### API Extensions
+
+None.
+
+### Risks and Mitigations
+
+The component combinations would need additional testing handled by
+an increased testing matrix.
+
+Considering networking/storage to be independent elements, probably
+an additional set of tests which cover "microshift-core" alone would
+be enough. When running those tests we need to disable PV and NetworkPolicy
+related testing.
+
+### Drawbacks
+
+TO-DO
+
+## Design Details
+
+### Open Questions
+
+1. How do we handle validation which is done today in go? (ovnk)
+   Would it make sense to have a validation.yaml that includes shell script
+   to be run on the system and validate anything necessary?, or should
+   we simply stop doing this type of validation?
+
+2. How do we handle CR upgrades when moving to newer versions of the APIs?,
+   See the `Upgrading`section.
+
+### Test Plan
+
+We will run additional test matrix with `microshift-core` alone, in those
+tests any OVN/NetworkPolicy tests would need to be disabled.
+
+Additional unit tests should cover the loading of core components as
+kustomize manifests, as well as the depencency handing between them.
+
+Additional unit tests should cover the loading of multiple `.d` style
+directories with manifests.
+
+## Implementation History
+
+* [openshift/microshift](https://github.com/openshift/microshift)
+* [Design guidelines](https://github.com/openshift/microshift/blob/main/docs/design.md)
+
+## Alternatives
+
+### Using go plugins which microshift can load when available, intead of yamls.
+
+[go/plugin](https://pkg.go.dev/plugin) could be used to create an interface
+for loadable components, MicroShift would scan an spefic directory for .so files
+
+Every plugin would contain it's own list of images, setup functionality,
+validation hooks, etc.
+
+The Risks here are:
+
+* Binary code duplication: How do we handle shared code between MicroShift
```

Comment:
This risk may not apply. Go plugin provides native golang interfaces for microshift to query manifests dir, renderParams etc, it doesn't have to call kustomize. The actual render/apply still happens in the microshift main process.

Reply:
But for the go plugin, could be. How would the plugin talk to the kubernetes API?, it needs the API client to talk to the different parts of kubernetes in it's own code binary

microshift can call the plugin,
the plugin can't call code living in microshift. (unless we provided callback functions but that would be overly twisted IMO)

Reply:
Go plugin doesn't have to talk to the kubernetes API, although it can if kubeconfig is provided.
In the proposal, go plugin interfaces are used to query the manifest path (installed by plugin package), plugin specific configurations (parsed by plugin code) etc.

Reply:
Ok, but if you pass kubeconfig and you need to talk to the API, the problem I describe would exist, as far as I understand.

Reply:
> Ok, but if you pass kubeconfig and you need to talk to the API, the problem I describe would exist, as far as I understand.

yes.